### PR TITLE
Fixing circular buffer.

### PR DESCRIPTION
**Pull request by VyacheslavIsaev**
Hi Mark, 

I am suggesting some fixes, which I came across.

1. Fixing ND circular buffer so it flushes pre-trigger buffer immediately on software trigger. It is because in my case there might not be a sequential image after the trigger.

2. Fixing NDArrayRing. Currently buffer holder is deleted on clear request. That makes reuse of the NDArrayRing impossible if you need to clear  NDArrayRing for example.

3. Small change in TIFF plugin which allows to read TIFFs generated in Photoshop or other editor which doesn't saves explicitly parameters with default value.
sampleFormat = SAMPLEFORMAT_UINT;

Thank you,
Slava Isaev

Project Leader
Cosylab d.d.

### CODE REVIEW

**Comment by MarkRivers** [comment]
This pull request is failing the Travis CI.  

```
1.65s$ ./bin/linux-x86_64/plugin-test --log_level=test_suite
Running 33 test cases...
Entering test suite "NDPlugin Tests"
Entering test suite "CircularBuffTests"
Entering test case "test_BufferWrappingAndStatusMessages"
Leaving test case "test_BufferWrappingAndStatusMessages"; testing time: 1984mks
Entering test case "test_OutputCount"
Leaving test case "test_OutputCount"; testing time: 749mks
Entering test case "test_PreBufferOrder"
../test_NDPluginCircularBuff.cpp(180): error in "test_PreBufferOrder": check 0 == ((uint8_t *)ds->arrays[0]->pData)[0] failed [0 != 0x3]
Leaving test case "test_PreBufferOrder"; testing time: 976mks
```

**Comment by MarkRivers** [comment]
I am assigning @ulrikpedersen to this since this plugin was written at Diamond by Edmund Warrick.

File: ADApp/pluginSrc/NDFileTIFF.cpp [hunk 1]
```diff
@@ -453,9 +462,11 @@ asynStatus NDFileTIFF::readFile(NDArray **pArray)
...
         return asynError;    
     }
-    if ((photoMetric == PHOTOMETRIC_MINISBLACK) && 
-        (planarConfig == PLANARCONFIG_CONTIG)   &&
-        (samplesPerPixel == 1)) {
+
+    if ((planarConfig == PLANARCONFIG_CONTIG)
+        && (samplesPerPixel == 1)
+		&& (photoMetric == PHOTOMETRIC_MINISBLACK))
+    {
```
[the remark this answers is not in the document]
Comment (VyacheslavIsaev) [reply ADApp/pluginSrc/NDFileTIFF.cpp]: Indeed, there is no difference at this part. Changes came out from my testing of different TIFF file formats. I have a problem with numerous TIFFs to load.

**Comment by VyacheslavIsaev** [comment]
Thank you for the review!

I decided to introduce a new parameter which will regulate flush behavior. 
By default flush will work as it was, but there will be an option to enable immediate flush on software trigger.

I will prepare new pull-request and will also take into account your comments!

**Pull request closed without merging**